[CHERIoT] Allow __cheri_libcall function pointers to decay to normal function pointers.

resistor opened the pull request:
https://github.com/CHERIoT-Platform/llvm-project/issues/16

resistor committed:
[CHERIoT] Allow __cheri_libcall function pointers to decay to normal function pointers.

v01dXYZ commented:
I know C++ is stricter with types but do you think the decay is still valid in this language too ? It doesn't seem this patch implements it for C++ (I tested your test case by renaming it to a cpp file).

resistor commented:
> I know C++ is stricter with types but do you think the decay is still valid in this language too ? It doesn't seem this patch implements it for C++ (I tested your test case by renaming it to a cpp file).

Good catch, thanks. Added a C++ test and fixed it. Unfortunately the C and C++ paths don't seem to coincide here.

davidchisnall reviewed (approved):
Looks good, thanks!
